### PR TITLE
Enable OS Updates to pre-release versions of higher base semver

### PR DESCRIPTION
**Pull request by thgreasi**
Depends-on: https://github.com/balena-io-modules/balena-hup-action-utils/pull/34
Depends-on: https://github.com/balena-io/balena-proxy/pull/1098
Update balena-hup-action-utils from 5.0.0 to 6.1.0

Change-type: minor

Resolves: # 
HQ: <url> 
See: <url> 
Depends-on: <url> 
Change-type: major|minor|patch 

---
##### Contributor checklist

- [ ] Includes tests
- [ ] Includes typings
- [ ] Includes updated documentation
- [ ] Includes updated build output

### CODE REVIEW

File: src/models/os.ts [hunk 1]
```diff
@@ -362,27 +362,32 @@ const getOsModel = function (
...
 			return await _getAllOsVersions(
 				deviceTypes,
-				listedByDefault
-					? {
+				filterOptions === 'all'
+					? undefined
+					: {
 							$filter: {
-								is_final: true,
+								...(filterOptions !== 'include_draft' && { is_final: true }),
```
Comment (otaviojacobi) [inline src/models/os.ts:375]: maybe filterOptions === 'default' is more clear?

File: src/models/os.ts [hunk 2]
```diff
@@ -362,27 +362,32 @@ const getOsModel = function (
 	};
 
 	const _memoizedGetAllOsVersions = authDependentMemoizer(
-		async (deviceTypes: string[], listedByDefault: boolean | null) => {
+		async (
+			deviceTypes: string[],
+			filterOptions: 'default' | 'include_draft' | 'all',
```
Comment (thgreasi) [inline src/models/os.ts:367]: `authDependentMemoizer` is atm using `primitive: true` and doesn't expose a normalizer, so I've left it for the next time we need to add an extra property for `getAvailableOsVersions`.

File: tests/integration/models/device.spec.ts [hunk 3]
```diff
@@ -3985,7 +3952,88 @@ describe('Device Model', function () {
...
+							expect(() =>
+								_checkOsUpdateTarget(
+									{
+										uuid,
+										is_of__device_type: [{ slug: 'raspberrypi3' }],
+										is_online: true,
+										os_version: osVersion,
+										os_variant: osVariant,
+									},
+									'2.29.2-1704382618288+rev1.prod',
```
Comment (myarmolinsky) [inline tests/integration/models/device.spec.ts:3992]: Can we please have comments on why these specific OS versions were chosen? Same on line 3999

File: tests/integration/models/device.spec.ts [hunk 4]
```diff
@@ -3985,7 +3952,88 @@ describe('Device Model', function () {
...
+							).to.throw('OS downgrades are not allowed');
+						});
+					});
+
+					it('should not throw when updating to a pre-release version with a newer base server', () => {
+						[['balenaOS 2.28.0+rev1', 'prod']].forEach(function ([
+							osVersion,
+							osVariant,
+						]) {
+							expect(() =>
```
Comment (myarmolinsky) [inline tests/integration/models/device.spec.ts:3983]: Why are we using an array and forEach when it's just one item?

File: tests/integration/models/device.spec.ts [hunk 5]
```diff
@@ -3985,7 +3952,88 @@ describe('Device Model', function () {
...
+							).to.not.throw();
+						});
+					});
+
+					it('should not throw when updating a device that is running a pre-release version to a version with a newer base server', () => {
+						[['balenaOS 2.28.0-1704382553234', 'prod']].forEach(function ([
+							osVersion,
+							osVariant,
+						]) {
+							expect(() =>
```
Comment (myarmolinsky) [inline tests/integration/models/device.spec.ts:4003]: Why are we using an array and forEach when it's just one item?

File: tests/integration/models/device.spec.ts [hunk 6]
```diff
@@ -3985,7 +3952,88 @@ describe('Device Model', function () {
...
+							).to.not.throw();
+						});
+					});
+
+					it('should not throw when updating a device that is running a pre-release version updating to a pre-release version with a newer base server', () => {
+						[['balenaOS 2.28.0-1704382553234', 'prod']].forEach(function ([
+							osVersion,
+							osVariant,
+						]) {
+							expect(() =>
```
Comment (myarmolinsky) [inline tests/integration/models/device.spec.ts:4023]: Why are we using an array and forEach when it's just one item?

File: src/models/os.ts [hunk 7]
```diff
@@ -404,13 +411,14 @@ const getOsModel = function (
 	 */
 	async function getAvailableOsVersions(
 		deviceTypes: string[] | string,
+		options?: { includeDraft?: boolean },
```
Comment (myarmolinsky) [inline src/models/os.ts:414]: Maybe this and line 900 should be the options that `_memoizedGetAllOsVersions` expects for its second parameter

File: typings/pinejs-client-core.d.ts
Comment (myarmolinsky) [inline typings/pinejs-client-core.d.ts:1]: Nothing changed? Just reformatted?

File: tests/integration/models/os.spec.ts [hunk 8]
```diff
@@ -304,6 +304,30 @@ describe('OS model', function () {
...
+					await balena.models.os.getAvailableOsVersions('raspberrypi3');
+				expect(versionInfos).to.be.an('array');
+
+				const draftVersions = versionInfos.filter(
+					(v) => bSemver.parse(v.raw_version)?.prerelease.length ?? 0 > 0,
+				);
+				expect(draftVersions).to.have.lengthOf(0);
+			});
+
+			it('should include draft OS versions when the respective flag is not used [string device type argument]', async () => {
```
Comment (myarmolinsky) [inline tests/integration/models/os.spec.ts:318]:
```suggestion
			it('should include draft OS versions when the respective flag is used [string device type argument]', async () => {
```

File: tests/integration/models/device.spec.ts (hunk 3 above)
Comment (thgreasi) [reply tests/integration/models/device.spec.ts:3992]: B/c that's the only one atm on prod that is draft :)

File: typings/pinejs-client-core.d.ts
Comment (thgreasi) [reply typings/pinejs-client-core.d.ts:1]: yes :)

File: src/models/os.ts (hunk 7 above)
Comment (thgreasi) [reply src/models/os.ts:414]: See: https://github.com/balena-io/balena-sdk/pull/1398#discussion_r1459184748

Comment (thgreasi) [reply src/models/os.ts:414]: Moreover `_memoizedGetAllOsVersions` is used by both `getAvailableOsVersions` & `getAllOsVersions`.
Till now I couldn't find a nice way to combine them :(

**Comment by myarmolinsky** [comment]
LGTM

File: tests/integration/models/device.spec.ts (hunk 3 above)
Comment (thgreasi) [reply tests/integration/models/device.spec.ts]: Ends up the versions in this file are random, so I added a comment about that right at the start of the `describe()`.

**Pull request merged**